opensrp-app/src/test/java/org/smartregister/login/interactor/BaseLoginInteractorTest.java

```diff
@@ -0,0 +1,403 @@
+
+    @Mock
+    private BaseLoginContract.Presenter presenter;
+
+    @Mock
+    private BaseLoginContract.View view;
+
+    @Mock
+    private Context context;
+
+    @Mock
+    private UserService userService;
+
+    @Mock
+    private AllSharedPreferences allSharedPreferences;
+
+    @Mock
+    private SyncConfiguration syncConfiguration;
+
+    @Captor
+    private ArgumentCaptor<DialogInterface.OnClickListener> dialogCaptor;
+
+    @Mock
+    private DialogInterface dialogInterface;
+
+    @Mock
+    private ResetAppHelper resetAppHelper;
+
+    @Captor
+    private ArgumentCaptor<OnCompleteClearDataCallback> onCompleteClearDataCaptor;
+
+    @Mock
+    private UniqueIdRepository uniqueIdRepository;
+
+    @Mock
+    private AllSettings allSettings;
+
+    @Captor
+    private ArgumentCaptor<Setting> settingCaptor;
+
+    private AppCompatActivity activity;
+
+    private LoginResponseData loginResponseData;
+
+    private String user = "johndoe";
+    private String password = "qwerty";
+
+    private UserService contextUserService;
+
+    @Before
+    public void setUp() {
+        contextUserService = CoreLibrary.getInstance().context().userService();
+        when(presenter.getOpenSRPContext()).thenReturn(context);
+        when(context.allSharedPreferences()).thenReturn(allSharedPreferences);
+        when(context.userService()).thenReturn(userService);
+        when(presenter.getLoginView()).thenReturn(view);
+        activity = Robolectric.buildActivity(AppCompatActivity.class).create().get();
+        when(view.getActivityContext()).thenReturn(activity);
+        loginResponseData = new LoginResponseData();
+        loginResponseData.user = new User().withUsername(user);
+        loginResponseData.time = new Time(new Date(), TimeZone.getTimeZone("Africa/Nairobi"));
+    }
+
+    @After
+    public void tearDown() {
+        Whitebox.setInternalState(CoreLibrary.getInstance().context(), "userService", contextUserService);
+    }
+
+    @Test
+    public void testOnDestroyShouldSetPresenterNull() {
+        assertNotNull(Whitebox.getInternalState(interactor, "mLoginPresenter"));
+        interactor.onDestroy(false);
+        assertNull(Whitebox.getInternalState(interactor, "mLoginPresenter"));
+    }
+
+    @Test
+    public void testLoginAttemptsRemoteLoginAndErrorsWithBaseURLIsMissing() {
+        when(allSharedPreferences.fetchBaseURL("")).thenReturn("");
+        interactor.login(new WeakReference<>(view), "johndoe", "password");
+        verify(view).hideKeyboard();
+        verify(view).enableLoginButton(false);
+        verify(allSharedPreferences).savePreference("DRISHTI_BASE_URL", activity.getString(R.string.opensrp_url));
+        verify(view).enableLoginButton(true);
+        verify(view).showErrorDialog(activity.getString(R.string.remote_login_base_url_missing_error));
+        verify(view, never()).goToHome(anyBoolean());
+    }
+
+    @Test
+    public void testLoginAttemptsRemoteLoginAndErrorsWithGenericError() {
```

Comment:
Its testing an exception block in `org.smartregister.login.interactor.BaseLoginInteractor#remoteLogin` this is because `getSharedPreferences().fetchBaseURL("").isEmpty() ` will throw an NPE since `getSharedPreferences` is mock

Reply:
Cool